plugin/plugins/filesystem/filesystem.go

```diff
@@ -0,0 +1,124 @@
+package filesystem
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/djdv/p9/p9"
+	plugin "github.com/ipfs/go-ipfs/plugin"
+	fsnodes "github.com/ipfs/go-ipfs/plugin/plugins/filesystem/nodes"
+	logging "github.com/ipfs/go-log"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+var (
+	_ plugin.PluginDaemon = (*FileSystemPlugin)(nil) // impl check
+
+	// Plugins is an exported list of plugins that will be loaded by go-ipfs.
+	Plugins = []plugin.Plugin{
+		&FileSystemPlugin{}, //TODO: individually name implementations: &P9{}
+	}
+
+	logger logging.EventLogger
+)
+
+func init() {
+	logger = logging.Logger("plugin/filesystem")
+}
+
+type FileSystemPlugin struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	addr      multiaddr.Multiaddr
+	listener  manet.Listener
+	errorChan chan error
+}
+
+func (*FileSystemPlugin) Name() string {
+	return PluginName
+}
+
+func (*FileSystemPlugin) Version() string {
+	return PluginVersion
+}
+
+func (fs *FileSystemPlugin) Init(env *plugin.Environment) error {
+	logger.Info("Initializing 9P resource server...")
+	if !filepath.IsAbs(env.Repo) {
+		absRepo, err := filepath.Abs(env.Repo)
+		if err != nil {
+			return err
+		}
+		env.Repo = absRepo
+	}
+
+	cfg := &Config{}
+	// config not being set is okay and will load defalts
+	// config being set with malformed data is not okay and will instruct the daemon to halt-and-catch-fire
+	rawConf, ok := (env.Config).(json.RawMessage)
+	if ok {
+		if err := json.Unmarshal(rawConf, cfg); err != nil {
+			return err
+		}
+	} else {
+		if env.Config != nil {
+			return fmt.Errorf("plugin config does not appear to be correctly formatted: %#v", env.Config)
+		}
+		cfg = defaultConfig(env.Repo)
+	}
+
+	var err error
+	if envAddr := os.ExpandEnv(EnvAddr); envAddr == "" {
```

Comment:
So this doesn't get lost, I'm still not a fan of using an environment variable like this. https://github.com/ipfs/go-ipfs/pull/6612#discussion_r322979599

Reply:
Resolved around here:
https://github.com/ipfs/go-ipfs/pull/6612/commits/21b2d2cf1b13db5076405173d02a803f4a8c241f#diff-1054eebd17f3591011a5e6aaa3b3bbdfR31

Reply:
>I'm still not a fan of using an environment variable like this.

Do the above changes seem better? I initially assumed "this" meant how the expansion was being processed. But I'm wondering now if you meant using env vars to override config file values, or maybe something else.

Reply:
@Stebalien ping on this comment ^